src/transport/AdminPairingTable.cpp

```diff
@@ -232,45 +244,105 @@ CHIP_ERROR AdminPairingInfo::SetRootCert(const ByteSpan & cert)
     }
-    mOpCertAllocatedLen = 0;
-    mOpCertLen          = 0;
-    mOperationalCert    = nullptr;
+    mICACertLen = 0;
+    mICACert    = nullptr;
 }
 
-CHIP_ERROR AdminPairingInfo::SetOperationalCert(const ByteSpan & cert)
+CHIP_ERROR AdminPairingInfo::SetICACert(const ByteSpan & cert)
 {
     if (cert.size() == 0)
     {
-        ReleaseOperationalCert();
+        ReleaseICACert();
         return CHIP_NO_ERROR;
     }
 
-    // There could be two certs in the set -> ICA and NOC
-    VerifyOrReturnError(cert.size() <= kMaxCHIPCertLength * 2, CHIP_ERROR_INVALID_ARGUMENT);
-    if (mOpCertLen != 0 && mOpCertAllocatedLen < cert.size())
+    VerifyOrReturnError(cert.size() <= kMaxCHIPCertLength, CHIP_ERROR_INVALID_ARGUMENT);
+    if (mICACertLen != 0)
+    {
+        ReleaseICACert();
+    }
+
+    VerifyOrReturnError(CanCastTo<uint16_t>(cert.size()), CHIP_ERROR_INVALID_ARGUMENT);
+    if (mICACert == nullptr)
+    {
+        mICACert = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(cert.size()));
+    }
+    VerifyOrReturnError(mICACert != nullptr, CHIP_ERROR_NO_MEMORY);
+    mICACertLen = static_cast<uint16_t>(cert.size());
+    memcpy(mICACert, cert.data(), mICACertLen);
+
+    return CHIP_NO_ERROR;
+}
+
+void AdminPairingInfo::ReleaseNOCCert()
+{
+    if (mNOCCert != nullptr)
+    {
+        chip::Platform::MemoryFree(mNOCCert);
+    }
+    mNOCCertLen = 0;
+    mNOCCert    = nullptr;
+}
+
+CHIP_ERROR AdminPairingInfo::SetNOCCert(const ByteSpan & cert)
+{
+    if (cert.size() == 0)
     {
-        ReleaseOperationalCert();
+        ReleaseNOCCert();
+        return CHIP_NO_ERROR;
     }
 
-    if (mOperationalCert == nullptr)
+    VerifyOrReturnError(cert.size() <= kMaxCHIPCertLength, CHIP_ERROR_INVALID_ARGUMENT);
+    if (mNOCCertLen != 0)
     {
-        mOperationalCert = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(cert.size()));
+        ReleaseNOCCert();
     }
-    VerifyOrReturnError(mOperationalCert != nullptr, CHIP_ERROR_NO_MEMORY);
+
     VerifyOrReturnError(CanCastTo<uint16_t>(cert.size()), CHIP_ERROR_INVALID_ARGUMENT);
-    mOpCertLen          = static_cast<uint16_t>(cert.size());
-    mOpCertAllocatedLen = (mOpCertLen > mOpCertAllocatedLen) ? mOpCertLen : mOpCertAllocatedLen;
-    memcpy(mOperationalCert, cert.data(), mOpCertLen);
+    if (mNOCCert == nullptr)
+    {
+        mNOCCert = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(cert.size()));
+    }
+    VerifyOrReturnError(mNOCCert != nullptr, CHIP_ERROR_NO_MEMORY);
+    mNOCCertLen = static_cast<uint16_t>(cert.size());
+    memcpy(mNOCCert, cert.data(), mNOCCertLen);
 
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR AdminPairingInfo::SetOperationalCertsFromCertArray(const ByteSpan & certArray)
+{
+    if (certArray.size() == 0)
+    {
+        ReleaseNOCCert();
+        ReleaseICACert();
+        return CHIP_NO_ERROR;
+    }
+
+    ByteSpan noc;
+    ByteSpan icac;
+    ExtractCertsFromCertArray(certArray, noc, icac);
```

Comment:
We don't need to split the cert array into NOC and ICAC. The CASESession is cable of loading the cert array itself using `LoadCerts` method. Besides that, is there any other reason for splitting it?

Reply:
The reason is that this avoids needing to store/allocate the contiguous block. No certificate chain should be stored as a linear chain. They appear on the wire as an array, but they don't need to follow through everywhere as an array. Keeping them separate allows eventually simplifying the storage arrangement to truly not require linear storage of ICAC + NOC.

Reply:
But if we always want them on the wire as a single blob, why would we not store them as that blob so we can blit directly onto the wire?

Otherwise we have to either transcode or do something where we store raw TLV bytes and insert them into a TLV we are constructing or something like that, right?